File: backend/src/main/java/com/taskflow/service/TareaService.java
package com.taskflow.service;

/*
 * Interfaz que define el contrato del servicio de tareas.
 * Aplica el principio de inversión de dependencias (DIP):
 * el controlador depende de una abstracción, no de una implementación.
 */
import com.taskflow.dto.TareaRequestDto;
import com.taskflow.dto.TareaResponseDto;
import com.taskflow.model.EstadoTarea;
import java.util.List;

/**
 * Interfaz del servicio de gestión de tareas.
 * Define todas las operaciones de negocio disponibles para las tareas.
 * 
 * @author TaskFlow Team
 * @version 1.0.0
 */
public interface TareaService {

    /**
     * Obtiene la lista completa de todas las tareas del sistema.
     * 
     * @return Lista de DTOs con todas las tareas
     */
    List<TareaResponseDto> obtenerTodasLasTareas();

    /**
     * Busca una tarea específica por su identificador único.
     * 
     * @param id Identificador de la tarea a buscar
     * @return DTO con los datos de la tarea encontrada
     * @throws com.taskflow.exception.RecursoNoEncontradoException si no existe
     */
    TareaResponseDto obtenerTareaPorId(Long id);

    /**
     * Crea una nueva tarea en el sistema con estado PENDIENTE.
     * 
     * @param dto DTO con los datos de la nueva tarea
     * @return DTO con la tarea creada (incluye ID generado)
     */
    TareaResponseDto crearTarea(TareaRequestDto dto);

    /**
     * Actualiza los datos de una tarea existente.
     * Solo modifica título, descripción y prioridad (no el estado).
     * 
     * @param id  Identificador de la tarea a actualizar
     * @param dto DTO con los nuevos datos
     * @return DTO con la tarea actualizada
     * @throws com.taskflow.exception.RecursoNoEncontradoException si no existe
     */
    TareaResponseDto actualizarTarea(Long id, TareaRequestDto dto);

    /**
     * Elimina una tarea del sistema por su identificador.
     * 
     * @param id Identificador de la tarea a eliminar
     * @throws com.taskflow.exception.RecursoNoEncontradoException si no existe
     */
    void eliminarTarea(Long id);

    /**
     * Cambia el estado de una tarea (ej: de PENDIENTE a EN_PROGRESO).
     * 
     * @param id     Identificador de la tarea
     * @param nuevoEstado Nuevo estado a asignar
     * @return DTO con la tarea actualizada
     * @throws com.taskflow.exception.RecursoNoEncontradoException si no existe
     */
    TareaResponseDto cambiarEstado(Long id, EstadoTarea nuevoEstado);

    /**
     * Filtra las tareas por un estado específico.
     * 
     * @param estado Estado por el cual filtrar
     * @return Lista de tareas que coinciden con el filtro
     */
    List<TareaResponseDto> filtrarPorEstado(EstadoTarea estado);

    /**
     * Busca tareas cuyo título contenga un texto dado.
     * 
     * @param texto Texto a buscar en los títulos
     * @return Lista de tareas que coinciden
     */
    List<TareaResponseDto> buscarPorTitulo(String texto);
}

File: frontend/src/services/TareaService.js
/**
 * Servicio de comunicación con la API REST del backend.
 * Centraliza todas las llamadas HTTP usando Axios.
 * La URL base se configura mediante el proxy de package.json.
 * 
 * @author TaskFlow Team
 * @version 1.0.0
 */
import axios from 'axios';

/** URL base de la API REST de tareas */
const API_URL = '/api/tareas';

/**
 * Obtiene tareas del backend con filtros opcionales.
 * 
 * @param {string} [estado=null] - Filtro por estado (PENDIENTE, EN_PROGRESO, etc.)
 * @param {string} [texto=null] - Texto de búsqueda en títulos
 * @returns {Promise<Array>} Lista de tareas obtenidas del backend
 */
export async function obtenerTareas(estado = null, texto = null) {
  try {
    let respuesta;
    if (texto) {
      // Usar endpoint de búsqueda por título
      respuesta = await axios.get(`${API_URL}/buscar`, {
        params: { texto }
      });
    } else if (estado && estado !== 'TODAS') {
      // Usar endpoint de filtro por estado
      respuesta = await axios.get(`${API_URL}/filtro`, {
        params: { estado }
      });
    } else {
      // Obtener todas las tareas sin filtro
      respuesta = await axios.get(API_URL);
    }
    return respuesta.data;
  } catch (error) {
    console.error('Error en obtenerTareas:', error);
    throw error;
  }
}

/**
 * Crea una nueva tarea en el backend.
 * 
 * @param {Object} tareaDatos - Datos de la tarea a crear
 * @param {string} tareaDatos.titulo - Título de la tarea (obligatorio)
 * @param {string} tareaDatos.descripcion - Descripción opcional
 * @param {number} tareaDatos.prioridad - Prioridad 1=Baja, 2=Media, 3=Alta
 * @returns {Promise<Object>} Tarea creada con ID generado
 */
export async function crearTarea(tareaDatos) {
  try {
    const respuesta = await axios.post(API_URL, tareaDatos);
    return respuesta.data;
  } catch (error) {
    console.error('Error en crearTarea:', error);
    throw error;
  }
}

/**
 * Actualiza una tarea existente en el backend.
 * 
 * @param {number} id - ID de la tarea a actualizar
 * @param {Object} tareaDatos - Nuevos datos para la tarea
 * @returns {Promise<Object>} Tarea actualizada
 */
export async function actualizarTarea(id, tareaDatos) {
  try {
    const respuesta = await axios.put(`${API_URL}/${id}`, tareaDatos);
    return respuesta.data;
  } catch (error) {
    console.error('Error en actualizarTarea:', error);
    throw error;
  }
}

/**
 * Elimina una tarea del backend por su ID.
 * 
 * @param {number} id - ID de la tarea a eliminar
 * @returns {Promise<void>} Respuesta vacía (204 No Content)
 */
export async function eliminarTarea(id) {
  try {
    await axios.delete(`${API_URL}/${id}`);
  } catch (error) {
    console.error('Error en eliminarTarea:', error);
    throw error;
  }
}

/**
 * Cambia el estado de una tarea.
 * Usa el endpoint PATCH para modificación parcial.
 * 
 * @param {number} id - ID de la tarea
 * @param {string} nuevoEstado - Nuevo estado (PENDIENTE, EN_PROGRESO, COMPLETADA, CANCELADA)
 * @returns {Promise<Object>} Tarea con el estado actualizado
 */
export async function cambiarEstadoTarea(id, nuevoEstado) {
  try {
    const respuesta = await axios.patch(`${API_URL}/${id}/estado`, null, {
      params: { nuevoEstado }
    });
    return respuesta.data;
  } catch (error) {
    console.error('Error en cambiarEstadoTarea:', error);
    throw error;
  }
}